tfhe/src/integer/server_key/radix_parallel/div_mod.rs

```diff
@@ -113,134 +114,241 @@ impl ServerKey {
+                self.key.generate_lookup_table_bivariate(|x, y| {
+                    u64::from(x == 0 && y == 0) << bit_position_in_block
+                })
+            })
+            .collect::<Vec<_>>();
+
         for i in (0..=total_bits as usize - 1).rev() {
-            let block_of_bit = i / num_bits_in_block as usize;
-            let pos_in_block = i % num_bits_in_block as usize;
+            let block_of_bit = i / num_bits_in_message as usize;
+            let pos_in_block = i % num_bits_in_message as usize;
 
-            // i goes from (total_bits - 1 to 0)
-            // msb_bit_set goes from 0 to total_bits - 1
+            // i goes from [total_bits - 1 to 0]
+            // msb_bit_set goes from [0 to total_bits - 1]
             let msb_bit_set = total_bits as usize - 1 - i;
-            let first_trivial_block = (msb_bit_set / num_bits_in_block as usize) + 1;
+
+            let last_non_trivial_block = msb_bit_set / num_bits_in_message as usize;
+            // Index to the first block of the remainder that is fully trivial 0
+            // and all blocks after it are also trivial zeros
+            // This number is in range 1..=num_bocks -1
+            let first_trivial_block = last_non_trivial_block + 1;
 
             // All blocks starting from the first_trivial_block are known to be trivial
             // So we can avoid work.
-            // Note that, these are always non-empty
+            // Note that, these are always non-emtpy (i.e there is always at least one non trivial
+            // block)
             let mut interesting_remainder =
-                RadixCiphertext::from(remainder.blocks[..first_trivial_block].to_vec());
+                RadixCiphertext::from(remainder.blocks[..=last_non_trivial_block].to_vec());
             let mut interesting_divisor =
-                RadixCiphertext::from(divisor.blocks[..first_trivial_block].to_vec());
+                RadixCiphertext::from(divisor.blocks[..=last_non_trivial_block].to_vec());
+            let mut divisor_ms_blocks = RadixCiphertext::from(
+                divisor.blocks[((msb_bit_set + 1) / num_bits_in_message as usize)..].to_vec(),
+            );
 
-            self.unchecked_scalar_left_shift_assign_parallelized(&mut interesting_remainder, 1);
-            self.key
-                .unchecked_add_assign(&mut interesting_remainder.blocks[0], &numerator_bits[i]);
+            // We split the divisor at a block position, when in reality the split should be at a
+            // bit position meaning that potentially (depending on msb_bit_set) the
+            // split versions share some bits they should not. So we do one PBS on the
+            // last block of the interesting_divisor, and first block of divisor_ms_blocks
+            // to trim out bits which should not be there
+
+            let mut trim_last_interesting_divisor_bits = || {
+                if ((msb_bit_set + 1) % num_bits_in_message as usize) == 0 {
```

Comment:
msb_bit_set + 1 seems to appear in several places, should it be stored in a variable with an explicit name ?

Reply:
I meant it may have some better human readable name if it does not, it's fine like this